_source/_posts/2022-03-15-programming-security-and-why-rust.md

```diff
@@ -0,0 +1,157 @@
+
+## Memory safety
+
+In a memory-safe language, when you access a variable or an item in an array, you can be sure that you are indeed accessing what you meant to or are allowed to access. In other words, you will not be reading or writing into the memory of another variable or pointer by mistake, regardless of what you do in your program.
+
+So, why is this a big deal? Don't all major programming languages ensure this?
+
+Yes, to varying extents. But some languages are unsafe by default—for example, C and C++. In C or C++, you can access the memory of another variable by mistake, or you can free a pointer twice; that's called double-free error. Sometimes a program continues to use a pointer after it has been freed, and that's called use after free errors or a dangling pointer error. Such behavior is categorized as undefined behavior as they are unpredictable and causes security vulnerability rather than just crashing the program. In this scenario, a crashing program is a good thing as it won't cause a security vulnerability.
+
+> I call it my billion-dollar mistake. It was the invention of the null reference in 1965
+>
+> - Tony Hoare
+
+Then there is also null safety which is kind of related to memory safety. I come from a Java/JavaScript background, and we are so used to the concept of null. Null is infamous for being the worst invention in programming. Garbage collected languages need a concept of nothing so that a pointer can be freed when unused. But it also leads to issues and pain, like the null pointer exceptions. Technically this falls under memory safety, but most memory-safe languages still let you use null as a value leading to null pointer errors.
+
+## Type safety
+
+In a type-safe language, when you access a variable, you access it as the correct type of data it is stored as. This gives us the confidence to work on data without manually checking for the data type during runtime. Memory safety is required for a language to be type-safe.
+
+## Thread safety
+
+In a thread-safe language, you can access or modify the same memory from multiple threads simultaneously without worrying about data races. This is generally achieved using message passing techniques, mutual exclusion locks (Mutex), and thread synchronization. Thread safety is required for optimal memory and type safety, so generally, memory and type-safe languages tend to be thread-safe as well.
+
+# Why does it matter?
+
+Ok! Why does this matter, and why should we care about it? Let's take a look at some stats to get an idea first.
+
+## Memory safety issues
+
+Memory safety issues are the cause of most security CVEs we encounter. Undefined behavior can be abused by a hacker to take control of the program or to leak privileged information. If you try to access an array element out of its bound in memory-safe languages, you will just crash the program with panic or error, which is predictable behavior.
```

Comment:
```suggestion
Memory safety issues are the cause of most security CVEs we encounter. Undefined behavior can be abused by a hacker to take control of the program or to leak privileged information. If you try to access an array element out of its bound in memory-safe languages, you will just crash the program with panic or error, which is predictable behavior.
```
```suggestion
Memory safety issues are the cause of most security CVEs (Common Vulnerabilities and Exposures) we encounter. Undefined behavior can be abused by a hacker to take control of the program or to leak privileged information. If you try to access an out of bounds array element in a memory-safe language, you will just crash the program with panic or error, which is predictable behavior.
```
spell out acronym once, align singular plural usage, use of "out of bounds" idiom

Reply:
I'm not sure why is appearing in duplicate, but the suggested change is as follows: 
Memory safety issues are the cause of most security CVEs (Common Vulnerabilities and Exposures) we encounter. Undefined behavior can be abused by a hacker to take control of the program or to leak privileged information. If you try to access an out of bounds array element in a memory-safe language, you will just crash the program with panic or error, which is predictable behavior.